packages/devextreme/testing/testcafe/helpers/generateOptionMatrix.ts

```diff
@@ -1,36 +1,70 @@
-export type Options<TComponentOptions = unknown> = {
+export type Options<TComponentOptions> = {
   [TProperty in keyof TComponentOptions]: TComponentOptions[TProperty][]
 };
 
-export const generateOptionMatrix = (
-  options: Options,
-  index = 0,
-  currentConfigurations: Options[] = [],
-): Options[] => {
-  const keys = Object.keys(options);
-
-  if (index === keys.length) {
-    return [Object.assign({}, ...currentConfigurations)];
+const generateOptionMatrixImpl = <T>(
+  options: Options<T>,
+  keys: (keyof T)[],
+  index: number,
+  current: T[],
+): T[] => {
+  if (index >= keys.length) {
+    return current;
   }
 
   const key = keys[index];
   const values = options[key];
+  const currentLength = current.length;
 
-  const configurations: Options[] = [];
+  for (let i = 1; i < values.length; i += 1) {
+    const nextCombinations = current
+      .slice(0, currentLength)
+      .map<T>((item) => ({
+      ...item,
+      [key]: values[i],
+    }));
 
-  values.forEach((value) => {
-    currentConfigurations.push({ [key]: value });
+    current.push(...nextCombinations);
+  }
 
-    const generatedConfigurations = generateOptionMatrix(
-      options,
-      index + 1,
-      currentConfigurations,
-    );
+  return generateOptionMatrixImpl(options, keys, index + 1, current);
+};
 
-    configurations.push(...generatedConfigurations);
+export const generateOptionMatrix = <T, U extends Partial<T> = Partial<T>>(
+  options: Options<T>,
+  exclude: U[] = [],
+): T[] => {
+  const keys = Object.keys(options) as (keyof T)[];
 
-    currentConfigurations.pop();
-  });
+  if (keys.length === 0) {
+    return [];
+  }
+
+  const initialObject = {} as T;
+
+  const baseObject = keys.reduce<T>((acc, key) => {
+    const [firstValue] = options[key];
+    acc[key] = firstValue;
+    return acc;
+  }, initialObject);
 
-  return configurations;
+  const current = options[keys[0]].map<T>((value) => ({
+    ...baseObject,
+    [keys[0]]: value,
+  }));
```

Comment:
Minor comment.

Let's use deconstruct to get the first value:
```suggestion
  const firstKey = [keys]
  const current = options[firstKey ].map<T>((value) => ({
    ...baseObject,
    [firstKey ]: value,
  }));
```